website/blog/2023-08-14-mongodb-sorting-scalar.md

```diff
@@ -0,0 +1,166 @@
+| 7                                           | ObjectId                                 |
+| 8                                           | Boolean                                  |
+| 9                                           | Date                                     |
+| 10                                          | Timestamp                                |
+| 11                                          | Regular Expression                       |
+
+## Comparison of different BSON types
+
+For comparing different BSON types, each BSON type has a predefined order of comparison assigned.
+For example, Null BSON type has the lowest order of comparison 1, so Null is less than any other BSON values.
+Boolean BSON type on the other hand has the order of comparison 8, a BSON type such as ObjectId with a lower order of comparison is less than a Boolean value
+and a BSON type such as Timestamp with a higher order of comparison is greater than a Boolean value.
+
+Comparing different BSON types is merely looking up the order of comparison table for each BSON type and comparing the predefined order.
+There is an exception for Array values, we will be discussing them in another blog post.
+
+### Number comparison
+
+Although numbers have different BSON types, namely Integer, Long, Double and Decimal, they are considered the equivalent BSON type for the purpose of comparison.
+That means comparing numbers consider their values but whether they are Integer, Long, Double or Decimal are not relevant.
+For instance, an Integer value 0 is equivalent to Double 0.0 as far as comparison is concerned.
+
+### Null and non-existent field comparison
+
+For the comparison purpose, non-existent field is equivalent to Null.
+This means that Null and non-existent field are equal as far as comparison is concerned.
+
+## Examples showcasing Sorting for scalar values
+
+Suppose there is an `outfits` collection, following query inserts documents.
+
+```js
+db.outfits.insertMany([
+  { _id: 1, name: 'flip flops', size: 'M', color: 'blue' },
+  { _id: 2, name: 'sandals', size: 9, color: null },
+  { _id: 3, name: 'boots', size: 8, color: 'black' },
+  { _id: 4, name: 'sneakers', size: 8.5, color: 'blue' },
+  { _id: 5, name: 'slippers' }
+])
+```
+
+The `outfits` collection has a `size` field, and it contains different BSON types.
+The document `flip flops` has String field value, `sandals` and `boots` have Integer field values,
+`sneakers` has Double field value and `slippers` is missing the field `size`.
+To sort the collection in ascending order by `size` field, sorting order 1 is used and the following query is run.
+
+```js
+db.outfits.find().sort({ size: 1 })
+```
+
+```json5
+[
+  { _id: 5, name: 'slippers' },
+  { _id: 3, name: 'boots', size: 8, color: 'black' },
+  { _id: 4, name: 'sneakers', size: 8.5, color: 'blue' },
+  { _id: 2, name: 'sandals', size: 9, color: null },
+  { _id: 1, name: 'flip flops', size: 'M', color: 'blue' }
+]
+```
```

Comment:
I like this example a lot! It really demonstrates the difference when we compare values of different types!